// @davdevs/paper-foundations/src/index.ts
export { Box } from './components/Box';
export type { BoxProps, BoxRef } from './components/Box';

export { Stack } from './components/Stack';
export type { StackProps, StackRef } from './components/Stack';

export { Inline } from './components/Inline';
export type { InlineProps, InlineRef } from './components/Inline';

export { Divider } from './components/Divider';
export type { DividerProps, DividerRef } from './components/Divider';

export { Surface } from './components/Surface';
export type { SurfaceProps, SurfaceRef } from './components/Surface';

export { Container } from './components/Container';
export type { ContainerProps, ContainerRef } from './components/Container';

export {
  Heading,
  Text,
  List,
  Blockquote,
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
  Lead,
  Small,
  Caption,
  Overline,
  Code,
} from './components/Typography';
export type {
  HeadingProps,
  TextProps,
  ListProps,
  BlockquoteProps,
  HeadingRef,
  TextRef,
  ListRef,
  BlockquoteRef,
} from './components/Typography';

// Export utilities
export { cn } from './utils';
